File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 012978. 배달/kotlin/Solution.kt
class Solution {
    fun solution(N: Int, road: Array<IntArray>, k: Int): Int {
        val inf = 20_000_000
        val graph = Array(N+1){ArrayList<Edge>()}
        val heap = Heap(road.size)
        val cost = IntArray(N+1) { inf }
        for (r in road) {
            graph[r[0]] += Edge(r[1], r[2])
            graph[r[1]] += Edge(r[0], r[2])
        }
        heap.push(Edge(1, 0))
        cost[1] = 0
        var cur: Edge
        var alt: Int
        var answer = 0
        while (heap.isNotEmpty()) {
            cur = heap.pop()
            if (cost[cur.node] < cur.cost) continue
            if (cur.cost <= k) answer ++
            for (edge in graph[cur.node]) {
                alt = cur.cost + edge.cost
                if (cost[edge.node] > alt) {
                    cost[edge.node] = alt
                    heap.push(Edge(edge.node, alt))
                }
            }
        }
        return answer
    }
}

class Edge(val node: Int, val cost: Int) : Comparable<Edge> {

    override fun compareTo(other: Edge): Int {
        return this.cost - other.cost
    }
}

class Heap(cap: Int) {

    private val arr = arrayOfNulls<Edge>(cap + 1)
    private var size = 0

    fun push(e: Edge) {
        arr[++size] = e
        var i = size
        var p = i shr 1
        while (p > 0) {
            if (arr[i]!! < arr[p]!!) {
                arr[i] = arr[p].also{arr[p] = arr[i]}
                i = p
                p = i shr 1
            } else {
                break
            }
        }
    }

    fun pop(): Edge {
        val e = arr[1]!!
        arr[1] = arr[size]
        arr[size--] = null

        var i = 1
        var lt = i shl 1
        var rt = lt + 1
        var small = i
        while (true) {
            if (lt <= size && arr[lt]!! < arr[small]!!) small = lt
            if (rt <= size && arr[rt]!! < arr[small]!!) small = rt
            if (i == small) break
            arr[i] = arr[small].also{arr[small] = arr[i]}
            i = small
            lt = i shl 1
            rt = lt + 1
        }
        return e
    }

    fun isNotEmpty() = size != 0
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 012978. 배달/python/solution.py
import heapq

def solution(N, road, K):

    # 그래프(인접 리스트)
    graph = {i: [] for i in range(1, N + 1)}
    for a, b, c in road:
        graph[a].append((b, c))
        graph[b].append((a, c))
    heap = []

    # 힙에는 (각 노드별 도달 비용, 노드 번호) 를 삽입
    heapq.heappush(heap, (0, 1))

    # 각 노드별 도달 최소 비용
    costs = [20_000_000] * (N + 1)
    costs[1] = 0
    answer = 0

    while heap:
        # 그 시점에서 노드 도달비용이 제일 작을 경우 추출됨
        cost, start = heapq.heappop(heap)

        if costs[start] < cost:
            continue

        # 이 시점에 왔을 때 그 노드에 도달한 최소비용 확정

        if costs[start] <= K:
            answer += 1

        # 인접한 노드들을 확인하며 최소 도달 비용을 갱신시키는 가능성이 있는 지 확인
        # 가능성이 있다 판단되면 최소비용 갱신
        # 힙에 삽입 (노드 도달비용, 노드)
        for end, next_cost in graph[start]:
            alt = cost + next_cost
            if costs[end] > alt:
                costs[end] = alt
                heapq.heappush(heap, (alt, end))

    return answer
